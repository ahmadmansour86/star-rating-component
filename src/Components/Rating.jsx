import React, { useState, useRef, useEffect } from 'react';
import Box from '@mui/material/Box';
import StarIcon from '@mui/icons-material/Star';
import Tooltip from '@mui/material/Tooltip';

const Rating = ({
  precision = 1,
  totalStars = 5,
  emptyIcon = StarIcon,
  filledIcon = StarIcon,
  fontSize = 'medium' || 'large'|| 'small',
  disabled = false,
  nonInteractive = false,
  defaultValue = 0, // Default rating value
  caption = false,
}) => {
  const [activeStar, setActiveStar] = useState(defaultValue);
  const [hoverActiveStar, setHoverActiveStar] = useState(-1);
  const [isHovered, setIsHovered] = useState(false);
  const ratingContainerRef = useRef(null);
  const disabledColor = '#eeeeee';

  const calculateRating = (e) => {
    const { width, left } = ratingContainerRef.current.getBoundingClientRect();
    let percent = (e.clientX - left) / width;
    const numberInStars = percent * totalStars;
    const nearestNumber = Math.round((numberInStars + precision / 2) / precision) * precision;

    return Number(nearestNumber.toFixed(precision.toString().split('.')[1]?.length || 0));
  };

  const handleClick = (e) => {
    if (!disabled && !nonInteractive) {
      setIsHovered(false);
      const clickedStar = calculateRating(e);
      if (clickedStar === activeStar) {
        setActiveStar(defaultValue);
      } else {
        setActiveStar(clickedStar);
      }
    }
  };

  const handleMouseMove = (e) => {
    if (!disabled && !nonInteractive) {
        setIsHovered(true);
        setHoverActiveStar(calculateRating(e));
      }
  };

  const handleMouseLeave = (e) => {
    if (!disabled && !nonInteractive) {
        setHoverActiveStar(-1); // Reset to default state
        setIsHovered(false);
      }
  };
  
  useEffect(() => {
    setActiveStar(defaultValue);
  }, [defaultValue]);

  const EmptyIcon = emptyIcon;
  const FilledIcon = filledIcon;

  return (
    <Box
      sx={{
        display: 'inline-flex',
        position: 'relative',
        cursor: disabled || nonInteractive ? 'auto' : 'pointer',
        textAlign: 'left',
        // Add focus style
        '&:focus': {
            outline: 'none',
            border: '2px solid #489edd',
            borderRadius: '10px'
          },
      }}
      onClick={handleClick}
      onMouseMove={handleMouseMove}
      onMouseLeave={handleMouseLeave}
      ref={ratingContainerRef}
    >
        
      {[...new Array(totalStars)].map((arr, index) => {
        const activeState = isHovered ? hoverActiveStar : activeStar;
        const showEmptyIcon = activeState === -1 || activeState < index + 1;
        const isActiveRating = activeState !== 1;
        const isRatingWithPrecision = activeState % 1 !== 0;
        const isRatingEqualToIndex = Math.ceil(activeState) === index + 1;
        const showRatingWithPrecision =
          isActiveRating && isRatingWithPrecision && isRatingEqualToIndex;
        const ratingTooltip = disabled || nonInteractive ? null : `${activeState} / ${totalStars}`;

        return (
        <Tooltip title={ratingTooltip} key={index} arrow={true} placement="top">
          <Box
            position={'relative'}
            sx={{
                cursor: disabled || nonInteractive ? 'auto' : 'pointer',
            }}
            key={index}
          >
            <Box
              sx={{
                width: showRatingWithPrecision ? `${(activeState % 1) * 100}%` : '0%',
                overflow: 'hidden',
                position: 'absolute',
                color: '#e8a93c'
              }}
            >
              <FilledIcon fontSize={fontSize} />
            </Box>
            
            <Box
              sx={{
                color: showEmptyIcon ? disabled ? disabledColor : '#e1e1e1' : disabled ? '#b8b8b7' : '#e8a93c'
              }}
            >
              {showEmptyIcon ? <EmptyIcon fontSize={fontSize} /> : <FilledIcon fontSize={fontSize}/>}
            </Box>
          </Box>
        </Tooltip>
        );
      })}
            <Box
                sx={{
                    display: caption ? 'inline-flex' : 'none',
                    position: 'relative',
                    padding: '4px 0 0 10px',
                }}>
                {activeStar}/{totalStars}
            </Box>
    </Box>
  );
};

export default Rating;