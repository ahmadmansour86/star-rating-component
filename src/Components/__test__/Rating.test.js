import React from "react";
import renderer from "react-test-renderer";
import Rating from '../Rating';


describe("Jest Snapshot testing suite", () => {
    it("Matches DOM Snapshot", () => {
      const domTree = renderer.create(<Rating />).toJSON();
      expect(domTree).toMatchSnapshot();
    });
  });
  
  describe("Jest Snapshot testing suite2", () => {
    it("Matches DOM Snapshot2", () => {
      const domTree = renderer.create(<Rating fontsize="small"/>).toJSON();
      expect(domTree).toMatchSnapshot();
    });
  });