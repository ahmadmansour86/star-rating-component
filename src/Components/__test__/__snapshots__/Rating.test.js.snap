// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Snapshot testing suite Matches DOM Snapshot 1`] = `
<div
  className="MuiBox-root css-1xivx0g"
  onClick={[Function]}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
>
  <div
    aria-label="0 / 5"
    aria-labelledby={null}
    className="MuiBox-root css-dfpqc0"
    data-mui-internal-clone-element={true}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="MuiBox-root css-1e3b1xn"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
    <div
      className="MuiBox-root css-1at5jfi"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
  </div>
  <div
    aria-label="0 / 5"
    aria-labelledby={null}
    className="MuiBox-root css-dfpqc0"
    data-mui-internal-clone-element={true}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="MuiBox-root css-1e3b1xn"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
    <div
      className="MuiBox-root css-1at5jfi"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
  </div>
  <div
    aria-label="0 / 5"
    aria-labelledby={null}
    className="MuiBox-root css-dfpqc0"
    data-mui-internal-clone-element={true}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="MuiBox-root css-1e3b1xn"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
    <div
      className="MuiBox-root css-1at5jfi"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
  </div>
  <div
    aria-label="0 / 5"
    aria-labelledby={null}
    className="MuiBox-root css-dfpqc0"
    data-mui-internal-clone-element={true}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="MuiBox-root css-1e3b1xn"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
    <div
      className="MuiBox-root css-1at5jfi"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
  </div>
  <div
    aria-label="0 / 5"
    aria-labelledby={null}
    className="MuiBox-root css-dfpqc0"
    data-mui-internal-clone-element={true}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="MuiBox-root css-1e3b1xn"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
    <div
      className="MuiBox-root css-1at5jfi"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
  </div>
  <div
    className="MuiBox-root css-1vp8k3g"
  >
    0
    /
    5
  </div>
</div>
`;

exports[`Jest Snapshot testing suite2 Matches DOM Snapshot2 1`] = `
<div
  className="MuiBox-root css-1xivx0g"
  onClick={[Function]}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
>
  <div
    aria-label="0 / 5"
    aria-labelledby={null}
    className="MuiBox-root css-dfpqc0"
    data-mui-internal-clone-element={true}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="MuiBox-root css-1e3b1xn"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
    <div
      className="MuiBox-root css-1at5jfi"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
  </div>
  <div
    aria-label="0 / 5"
    aria-labelledby={null}
    className="MuiBox-root css-dfpqc0"
    data-mui-internal-clone-element={true}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="MuiBox-root css-1e3b1xn"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
    <div
      className="MuiBox-root css-1at5jfi"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
  </div>
  <div
    aria-label="0 / 5"
    aria-labelledby={null}
    className="MuiBox-root css-dfpqc0"
    data-mui-internal-clone-element={true}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="MuiBox-root css-1e3b1xn"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
    <div
      className="MuiBox-root css-1at5jfi"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
  </div>
  <div
    aria-label="0 / 5"
    aria-labelledby={null}
    className="MuiBox-root css-dfpqc0"
    data-mui-internal-clone-element={true}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="MuiBox-root css-1e3b1xn"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
    <div
      className="MuiBox-root css-1at5jfi"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
  </div>
  <div
    aria-label="0 / 5"
    aria-labelledby={null}
    className="MuiBox-root css-dfpqc0"
    data-mui-internal-clone-element={true}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="MuiBox-root css-1e3b1xn"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
    <div
      className="MuiBox-root css-1at5jfi"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </div>
  </div>
  <div
    className="MuiBox-root css-1vp8k3g"
  >
    0
    /
    5
  </div>
</div>
`;
